scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/mirrors/OfflineMirror.scala

```diff
@@ -0,0 +1,56 @@
+package scala.meta
+package internal
+package semantic
+package mirrors
+
+import java.io._
+import org.scalameta.data._
+import scala.{Seq => _}
+import scala.collection.immutable.Seq
+import scala.collection.mutable
+import scala.meta.internal.io.PlatformIO
+import scala.meta.parsers._
+import scala.meta.semantic._
+
+@data class OfflineMirror(classpath: String, sourcepath: String) extends CommonMirror {
+  private def failEmpty(what: String) =
+    sys.error(
+      s"$what must be non-empty. " +
+        s"This may indicate that Mirror is badly configured. " +
+        s"If you use sbt-scalahost, make sure your project defines " +
+        s"`dependsOn(<projectname> % Scalameta)` for at least one <projectname>.")
+  if (classpath == null || classpath == "") failEmpty("classpath")
```

Comment:
That's an interesting way of looking at this. I think I agree that involved classes like that shouldn't be `@data`.